src/commonMain/kotlin/exchange.dydx.abacus/utils/AnalyticsUtils.kt

```diff
@@ -1,26 +1,97 @@
 import exchange.dydx.abacus.state.manager.HumanReadableCancelOrderPayload
 import exchange.dydx.abacus.state.manager.HumanReadablePlaceOrderPayload
+import exchange.dydx.abacus.state.manager.HumanReadableTriggerOrdersPayload
+import kollections.JsExport
 import kollections.toIMap
+import kotlinx.serialization.Serializable
+
+@JsExport
+@Serializable
+enum class TriggerOrderAction(val rawValue: String) {
+    REPLACE("REPLACE"),
+    CANCEL("CANCEL"),
+    CREATE("CREATE"),
+    ;
+
+    companion object {
+        operator fun invoke(rawValue: String?) =
+            TriggerOrderAction.values().firstOrNull { it.rawValue == rawValue }
+    }
+}
 
 class AnalyticsUtils {
+    /**
+     * Format Trigger Orders Payload and add additional details for `TriggerOrders` Analytic Events
+     * @param payload HumanReadableTriggerOrdersPayload
+     */
+
+    /**
+     * Format Place Order Payload for `TriggerOrders` Analytic Event
+     * @param payload HumanReadableTriggerOrdersPayload
+     */
+    fun triggerOrdersAnalyticsPayload(
+        payload: HumanReadableTriggerOrdersPayload,
+    ): IMap<String, Any>? {
+        val placeOrderPayloads = payload.placeOrderPayloads
+        val cancelOrderPayloads = payload.cancelOrderPayloads
+
+        val stopLossOrderTypes = listOf(OrderType.stopMarket, OrderType.stopLimit)
+        val takeProfitOrderTypes = listOf(OrderType.takeProfitMarket, OrderType.takeProfitLimit)
+
+        var stopLossOrderAction: TriggerOrderAction? = null
+        var takeProfitOrderAction: TriggerOrderAction? = null
+
+        placeOrderPayloads.forEach { placePayload ->
+            val orderType = OrderType.invoke(placePayload.type)
+            if (stopLossOrderTypes.contains(orderType)) {
+                stopLossOrderAction = TriggerOrderAction.CREATE
+            } else if (takeProfitOrderTypes.contains(orderType)) {
+                takeProfitOrderAction = TriggerOrderAction.CREATE
+            }
+        }
+
+        cancelOrderPayloads.forEach { cancelPayload ->
+            val orderType = OrderType.invoke(cancelPayload.type)
+            if (stopLossOrderTypes.contains(orderType)) {
+                stopLossOrderAction = if (stopLossOrderAction == null) {
+                    TriggerOrderAction.CANCEL
+                } else {
+                    TriggerOrderAction.REPLACE
+                }
+            } else if (takeProfitOrderTypes.contains(orderType)) {
+                takeProfitOrderAction = if (takeProfitOrderAction == null) {
+                    TriggerOrderAction.CANCEL
+                } else {
+                    TriggerOrderAction.REPLACE
+                }
+            }
+        }
+
+        return iMapOf(
+            "marketId" to payload.marketId,
+            "positionSize" to payload.positionSize,
+            "stopLossOrderAction" to stopLossOrderAction?.rawValue,
+            "takeProfitOrderAction" to takeProfitOrderAction?.rawValue,
+        ) as IMap<String, Any>?
```

Comment:
ah good point - should be fine if I just update the return type to be `IMap<String, Any?>`

Reply:
oh actually I see why we have all these casts in the file - it's because `ParsingHelper.merge` expects and returns `Map<String, Any>?` instead of `Map<String, Any?>`. Given how widely this is used, it seems like a pretty beefy change to fix, so I might just keep it like this for now 😢